[Wallet] Run E2E test on iOS

jeanregisser opened the pull request:
### Description

This PR enables the wallet E2E test on iOS and runs it on CircleCI.

Main highlights:
- Spent a lot of time making sure the test runs reliably. Let's hope it will pass the test of time 🤞 🍀 
- Made detox initialization failures easier to spot and troubleshoot. Previously it would still try to run the test suite after an init failure, leading to a lot of confusing errors. Detox init now blocks further progress if it fails, example: <img width="1164" alt="Screenshot 2020-02-11 at 14 02 25" src="https://user-images.githubusercontent.com/57791/74255336-005b4700-4cf2-11ea-8876-470454daaeb7.png">
- `victory` charts package was preventing the E2E test to run because it's using `d3-timer` under the hood which itself uses `requestAnimationFrame` continuously. Detox itself monitors all timers shorter than 1.5 secs and doesn't mark itself as ready while those are still running. I had to monkey patch it to workaround this during the E2E test
- We're now preloading the app bundle before starting the E2E test. Without this, the app was randomly failing to start fully with the red screen `Could not connect to development server`, though I could see the bundle building reach 100% in the packager logs when this happened.
- Improved `yarn install` time on CircleCI.
  - Stats for the `install_dependencies` step:
    - previously it took ~9mins on CircleCI (with or without an existing cache)
    - now with an empty cache it still takes ~9 mins, but with a populate cached it's now around 5 mins.
- Reported the issue of `victory-native` including `victory-core` twice: https://github.com/FormidableLabs/victory-native/issues/535
- CocoaPods now uses the CDN for faster installs. Especially true on CircleCI as the huge CocoaPods specs git repo doesn't need to be cloned anymore. See http://blog.cocoapods.org/CocoaPods-1.7.2/
- Fail builds if `Podfile.lock` or `Gemfile.lock` have differences after the install step. This is to help developers remember to run the install commands when touching the dependencies. As is already the case for yarn.
- Skip uploading iOS debug symbols to Sentry for debug builds

Future work:
- restore running E2E test on Android on CircleCI
- update skipped or outdated E2E tests (right now it doesn't do much, launches the app and goes through the first few screens only. i.e. no actual transactions)

### Tested

iOS E2E test runs successfully on CircleCI.

### Other changes

- Improved `stringToBoolean` util to trim whitespace.
- Removed now unneeded `npm rebuild scrypt` step in CircleCI

### Related issues

- Epic https://github.com/celo-org/celo-monorepo/issues/2641
- #1659
- #1154

### Backwards compatibility

Yes.

jeanregisser committed:
Make detox e2e tests work on iOS

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/ci-build-ios

jeanregisser committed:
Workaround victory-native importing 2 different victory-core modules

jeanregisser committed:
Made stringToBoolean parsing accept start or end whitespace

jeanregisser committed:
Skip firebase configure when disabled

jeanregisser committed:
Use pipefail to abort test if compilation fails

jeanregisser committed:
Add ios e2e job

jeanregisser committed:
Install needed tools and use a different checkout

jeanregisser committed:
Build utils (required)

jeanregisser committed:
Define and use common yarn_install command

jeanregisser committed:
Use version 2.1 to be able to use custom commands

jeanregisser committed:
Remove generic fallback as it doesn't work between executors

jeanregisser committed:
No need to create an empty watchman config

jeanregisser committed:
Switch to cocoapods cdn for faster install

Especially true on the CI as the huge cocoapods specs repo doesn't need to be cloned anymore with this.

jeanregisser committed:
Cache pods

jeanregisser committed:
Add template for sentry config for CI builds

jeanregisser committed:
Skip uploading debug symbols for debug builds

jeanregisser committed:
Build all modile deps

jeanregisser committed:
Fix build command

jeanregisser committed:
Temp debug yarn cache not working

jeanregisser committed:
Workaround save_cache not supporting globbing paths

jeanregisser committed:
Bump cache version

jeanregisser committed:
Remove deletion of node_modules/@celo

This shouldn't be needed anymore

jeanregisser committed:
Fix detox framework not present when restoring node_modules from cache

jeanregisser committed:
Re-enable install_dependencies

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/ci-build-ios

jeanregisser committed:
Improve node_modules cache

- use yarn.lock and patches content in the cache key
- create cache only when necessary

jeanregisser committed:
Speed up node_modules.tgz creation

jeanregisser committed:
Cache yarn global cache

jeanregisser committed:
Make sure the concatenated patches file remains stable no matter what the find iteration does

jeanregisser committed:
Don't use the branch for the yarn global cache key

jeanregisser committed:
Empty commit to test cache

jeanregisser committed:
Remove caching yarn global cache as it's too big and don't end up being a net positive

jeanregisser committed:
Remove rebuild scrypt steps as it's not needed anymore

jeanregisser committed:
Increase init timeout to see if it helps with the occasional issue seen in CircleCI

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/ci-build-ios

jeanregisser committed:
Skip licenses check when deps haven't changed

jeanregisser committed:
Prevent random red screen on CI

jeanregisser committed:
Cache built detox framewrok

jeanregisser committed:
Remove max-workers=1 option as it's too slow and runs out of memory on the CI

jeanregisser committed:
Tweak echo

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/ci-build-ios

jeanregisser committed:
Fail if lockfile changed after running pod install / bundle install

jeanregisser committed:
Fix specs repo

jeanregisser committed:
Make Detox init issues easier to throubleshoot

jeanregisser committed:
Improve troubleshooting steps

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/ci-build-ios

jeanregisser committed:
Add meaningful command description

jeanregisser committed:
Re-enable other steps

jeanregisser committed:
Typo

jeanregisser committed:
Link issue about multiple versions of victory-core included by victory-native

jmrossy commented on packages/mobile/e2e/Transfer.spec.js:
```diff
@@ -56,7 +56,13 @@ describe('Transfer Works', () => {
-    enterPin()
+
+    if (device.getPlatform() === 'android') {
+      enterPin()
```
Android should use the custom keypad too. I removed the scripts bits that unlocked the android emulator to improve the reliability

jeanregisser replied:
Thanks will update this! I wasn't able to run the Android test as mentioned on Slack.

jmrossy commented on packages/mobile/src/setupE2eEnv.e2e.ts:
```diff
@@ -0,0 +1,26 @@
+import { TimerContext as TimerContextES } from 'victory-core/es'
+
+// Inject custom e2e config
+Config.IS_E2E = true
```
Why inject this in here instead of setting `IS_E2E=true` in ` .env.test`? I thought I added it there already, I think that was my mistake

jeanregisser replied:
There was a script adding this variable to the env file, but it was removed recently and IS_E2E wasn't set anywhere anymore.
I thought about adding it back to the env file, but then I found that it was more flexible to have it injected this way as it allows using any existing env file without modifying it.

jmrossy commented on packages/mobile/src/setupE2eEnv.e2e.ts:
```diff
@@ -0,0 +1,26 @@
+import Config from 'react-native-config'
```
What causes this `.e2e.ts` file to be imported instead of the other ts file? Is that a Detox thing?

jeanregisser replied:
It's activated in `metro.config.js` with the env var `CELO_TEST_CONFIG=e2e`.

jmrossy reviewed (commented):
Amazing work! Just a few small questions below

i1skn reviewed (approved):
Thanks for the great effort on making this work! LGTM!

jeanregisser committed:
Add comment about why IS_E2E var is injected instead of set in an env file

jeanregisser committed:
Both iOS and Android should use the custom keypad

jeanregisser committed:
Keep e2e artifacts

- Factorize e2e run options for iOS and Android
- Always generate screenshots
- store artifacts in CI

jeanregisser committed:
Don't do partial cache restoration for pods

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/ci-build-ios

jeanregisser committed:
Merge branch 'master' into jeanregisser/ci-build-ios

jeanregisser committed:
Merge branch 'master' into jeanregisser/ci-build-ios
